Recommend the usage of the `to` header for signed messages (#296)
decentralized-identity/didcomm-messaging: docs/spec-files/message_structure.md

**DIDComm signed messages** are not necessary to provide message integrity (tamper evidence), or to prove the sender to the recipient. Both of these guarantees automatically occur with the authenticated encryption in **DIDComm encrypted messages**. **DIDComm signed messages** are only necessary when the origin of plaintext has to be provable to third parties, or when the sender can't be proven to the recipient by authenticated encryption because the recipient is not known in advance (e.g., in a broadcast scenario). Adding a signature when one is not needed [can degrade rather than enhance security because it relinquishes the sender's ability to speak off the record](https://github.com/hyperledger/aries-rfcs/blob/master/concepts/0049-repudiation/README.md#summary). We therefore expect **DIDComm signed messages** to be used in a few cases, but not as a matter of course.

> When a message is *both* signed and encrypted, the plaintext SHOULD be signed, and then the signed envelope SHOULD be encrypted. The opposite order SHOULD NOT be used, since it would imply that the signer committed to opaque data (which is unsafe and undermines non-repudiation).

Reviewer: ```suggestion
When a message is *both* signed and encrypted, the plaintext MUST be signed, and then the signed envelope MUST be encrypted. The opposite order MUST NOT be used, since it would imply that the signer committed to opaque data (which is unsafe and undermines non-repudiation).
```

I see no justifiable reason to do the opposite ordering, so I'd prefer to cut that code path out with a MUST here.

Author: I could see protocols where you have a signature on a ciphertext, in order to allow anyone to verify the provenance of the ciphertext without allowing decryption...

Typically you could do Sign then Encrypt then Mac/Sign... Having a strong requirement here seemed a bit restrictive. 

Also anything that is based on homomorphic encryption might want to not embed the signature within its ciphertext... Arguably, the structure of the plaintext wouldn't allow for such schemes with DIDComm, so it's probably not a problem to rule out such possibilities.

Reviewer: I don't see value in adding that additional code path here. I'd prefer to leave these problems out of scope here since it goes against the [recommendations of JOSE](https://datatracker.ietf.org/doc/html/rfc7519#section-11.2). I'd prefer to remain in alignment with this rather than support these additional aspects being suggested here.

Author: How about not using normative terminology and linking to that JOSE section?

Author: @kdenhartog ?

The [media type](https://tools.ietf.org/html/rfc6838) of a **DIDComm signed message** MUST be `application/didcomm-signed+json`.

The media type of the envelope MUST be set in the `typ` [property](https://tools.ietf.org/html/rfc7515#section-4.1.9) of the JWS.

> In order to avoid [surreptitious forwarding or malicious usage](https://theworld.com/~dtd/sign_encrypt/sign_encrypt7.html) of a signed message, a **DIDComm signed message** SHOULD contain a properly defined `to` header, this is also the case when a message is *both* signed and encrypted.

Reviewer: ```suggestion
In order to avoid [surreptitious forwarding or malicious usage](https://theworld.com/~dtd/sign_encrypt/sign_encrypt7.html) of a signed message, a **DIDComm signed message** SHOULD contain a properly defined `to` header, In the case when a message is *both* signed and encrypted the inner JWM being signed MUST contain a `to` header.
```

> In order to avoid [surreptitious forwarding or malicious usage](https://theworld.com/~dtd/sign_encrypt/sign_encrypt7.html) of a signed message, a **DIDComm signed message** SHOULD contain a properly defined `to` header. In the case where a message is *both* signed and encrypted, the inner JWM being signed MUST contain a `to` header.

Author: @kdenhartog Notice that here we are recommending (SHOULD) usage of the header for "just" signed messages, but mandating (MUST) it for signed and encrypted message, as per your suggested change. That's fine for me, but I just wanted to double check with you that's what you intended.

Reviewer: Yeah that's what I was thinking for this. This looks good.

- **type** - REQUIRED. Plaintext message type, useful for message handling in application-level protocols. The `type` attribute value MUST be a valid [Message Type URI](protocols.md#message-type-uri), that when resolved gives human readable information about the message. The attribute's value SHOULD predict the content in the `body` of the message.

- **typ** - OPTIONAL. Media type of the JWM content.

> - **to** - OPTIONAL. Identifier(s) for recipients. MUST be an array of strings where each element is a valid [DID](https://www.w3.org/TR/did-core/) or [DID URL](https://w3c.github.io/did-core/#did-url-syntax) (without the [fragment component](https://w3c.github.io/did-core/#fragment)) that identifies a member of the message's intended audience.

Reviewer: I'd prefer to see us move the test into this section here to remain consistent in our definition of headers rather than move the normative requirements into a separate section which doesn't call out that it's making requirements around the usage of the `to` header.

Author: This is just removing the duplicated entry for the "to" header. This was duplicated in #277 without reason.

I think it was a copy pasta mistake.

If you look below in that file, it's still containing this exact same paragraph.

Author: It seems @TelegramSam is also fixing this in his latest PR #274.
So I guess I can drop this removal from this PR.

Reviewer: Thanks for pointing that out, I didn't realize this.

 When Alice sends the same message to Bob and Carol, it is by inspecting this header that Bob and Carol learn that the message was sent to both of them. If the header is omitted, each recipient can only assume they are the only recipient (much like an email sent only to `BCC:` addresses).

> For signed messages, notice that it is RECOMMENDED to have a properly defined `to` header, since it prevents certain kind of [forwarding attacks](https://theworld.com/~dtd/sign_encrypt/sign_encrypt7.html), where a message that wasn't meant for a given recipient is forwarded along with its signature to that recipient which then could blindly trust it because of the signature.

Reviewer: ```suggestion
For signed messages, there's specific requirements around properly defined `to` header outlined in the `to` header definition above. The reason for this is to prevents certain kind of [forwarding attacks](https://theworld.com/~dtd/sign_encrypt/sign_encrypt7.html), where a message that wasn't meant for a given recipient is forwarded along with its signature to that recipient which then could blindly trust it because of the signature.
```

Reviewer: Changing this so that someone reading this section knows to reference the other section for requirements so we're not redundantly stating normative statements.

> Upon reception of a message whose `to` header is defined, the recipient SHOULD verify that they are included in that field and implementations SHOULD give a warning to their user when it is not the case as it could indicate malicious intent from the sender.

Reviewer: I'm good with these SHOULDs here
